src/Core/Ripgrep.re

```diff
@@ -14,7 +32,60 @@ let getCompletedCount = () => _ripGrepCompletedCount^;
 type searchFunction =
   (string, list(string) => unit, unit => unit) => disposeFunction;
 
-type t = {search: searchFunction};
+module Match = {
+  type t = {
+    file: string,
+    text: string,
+    lineNumber: int,
+    charStart: int,
+    charEnd: int,
+  };
+
+  let fromJsonString = str => {
+    Yojson.Basic.Util.(
+      try({
+        let json = Yojson.Basic.from_string(str);
+
+        if (member("type", json) == `String("match")) {
+          let data = member("data", json);
+
+          let submatches = data |> member("submatches") |> to_list;
+
+          let matches =
+            submatches
+            |> List.map(submatch =>
+                 {
+                   file:
+                     data |> member("path") |> member("text") |> to_string,
```

Comment:
Cool! I think I've duplicated some of these Util methods like `to_string` before - good to know for the future.

Reply:
I'm more fond of the more declarative combinator-style API of Elm's ['Json.Decode'](https://package.elm-lang.org/packages/elm/json/1.1.3/) or [`bs-json`](https://github.com/glennsl/bs-json), but this isn't too far away from that.

It's a nice API and it's usually better to decode explicitly like this than to use 'deriving', I think, since this decodes into exactly what you need rather than into some intermediate structure that just ends up being more boilerplate. Or worse, you end up molding your application to fit external data structures rather than using the data structures that are appropriate for your application needs, which tends to lead to all sorts of trouble later.